src/front/wgsl/mod.rs

```diff
@@ -2158,96 +2224,134 @@ impl Parser {
                         }?;
 
                         crate::Expression::AccessIndex {
-                            base: handle,
+                            base: expr.handle,
                             index,
                         }
                     } else {
                         crate::Expression::Access {
-                            base: handle,
+                            base: expr.handle,
                             index,
                         }
                     }
                 }
                 _ => {
-                    // after we reached for the value, load it
-                    return Ok(if needs_deref {
-                        let expression = crate::Expression::Load { pointer: handle };
-                        ctx.expressions
-                            .append(expression, ctx.expressions.get_span(handle))
-                    } else {
-                        handle
-                    });
+                    return Ok(expr);
                 }
             };
 
-            handle = ctx
+            expr.handle = ctx
                 .expressions
                 .append(expression, NagaSpan::from(lexer.span_from(span_start)));
         }
     }
 
-    fn parse_singular_expression<'a>(
+    /// Parse a `unary_expression`.
+    fn parse_unary_expression<'a>(
         &mut self,
         lexer: &mut Lexer<'a>,
         mut ctx: ExpressionContext<'a, '_, '_>,
-    ) -> Result<Handle<crate::Expression>, Error<'a>> {
-        let start = lexer.current_byte_offset();
-        self.push_scope(Scope::SingularExpr, lexer);
+    ) -> Result<TypedExpression, Error<'a>> {
+        self.push_scope(Scope::UnaryExpr, lexer);
         //TODO: refactor this to avoid backing up
-        let (allow_deref, handle) = match lexer.peek().0 {
+        let expr = match lexer.peek().0 {
             Token::Operation('-') => {
                 let _ = lexer.next();
+                let unloaded_expr = self.parse_unary_expression(lexer, ctx.reborrow())?;
```

Comment:
the "unloaded" prefix is confusing me. It sounds like something that was *done* to an expression, like "evicted". But that's not what's going on here. It's difficult to name properly. One candidate is to name it according to the type, i.e. "typed_expr", although it's definitely not the best either.

Reply:
It's only used in one place, so `parse_general_expression_for_reference` might be tolerable.